### DISCUSSION TITLE
Use PnPjs in a custom web component

### DISCUSSION THREAD

**Discussion by MarkusLanger** [microsoft-search/pnp-modern-search#2547] [Q&A] [answered]
**Is your feature request related to a problem? Please describe.**
I want to create a [custom web component](https://microsoft-search.github.io/pnp-modern-search/extensibility/custom_web_component/) and use PnPjs to receive data. The [initialization](https://pnp.github.io/pnpjs/getting-started/#using-pnpsp-spfi-factory-interface-in-spfx) of PnPjs requires the current context.

**Describe the solution you'd like**
How is it possible to initialize PnPjs in a custom web component? I found several issues but without an idea to solve my request.

- [Private  property this._ctx](https://github.com/microsoft-search/pnp-modern-search/issues/52) is not working anymore 
- [MSGraph from serviceScope](https://github.com/microsoft-search/pnp-modern-search/issues/918)

**Additional context**
- @pnp/modern-search-extensibility is version 1.4.15
- My library component has the same SPFx Version 1.14.0 as PnP Modern Search

**Comment by patrikhellgren** [comment, marked as answer]
See #865 on the way to get the page context.
Then you can use the following to initialize a PnPjs Web object in the react component used by the custom web component:
```
const web = Web(item["SPSiteUrl"]).using(SPFx({pageContext: props.pageContext})).using(Caching());
```

This is of course just one way of initializing PnPjs, have a look in the documentation for the others.

**Reply by Maya-Mostafa** [reply]
@patrikhellgren, I tried using this method with no luck. I really appreciate your help here. I was successful in getting the page context from the service scope but not the web part context. 

`const web = Web("https://mySPSiteUrl").using(SPFx({pageContext: props.pageContext})).using(Caching());`
When I logged this constant, it is returning a Proxy function, how could I handle this and pass it to another component? What am I missing here?

I am using the People Picker (PnP Js component) which requires the web part context to be passed to it, so of course, I won't be able to get this working unless I get the current context.

Thanks!

**Reply by patrikhellgren** [reply]
@Maya-Mostafa Not sure what you are trying to accomplish but web in this case has nothing to do with web part context.

If you want to pass BaseComponentContext to use in PeoplePicker in a Modern Search Custom Web Component you can accomplish this by constructing your own context instance like this in the web component (with irrelevant code stripped):

``` TypeScript
    interface IExampleProps {
        context?: any;
    }

    public async connectedCallback() {
        const props = this.resolveAttributes() as IExampleProps;
        props.context = {}
        props.context.serviceScope = this._serviceScope;
        props.context.pageContext = this._serviceScope.consume(PageContext.serviceKey);
        props.context.spHttpClient = this._serviceScope.consume(SPHttpClient.serviceKey);  
        props.context.aadTokenProviderFactory = this._serviceScope.consume(AadTokenProviderFactory.serviceKey);
        props.context.msGraphClientFactory = this._serviceScope.consume(MSGraphClientFactory.serviceKey);
        const component = <ExampleReactComponent {...props} />;
        ReactDom.render(component, this);
    }

```
And then in ExampleReactComponent:
``` TypeScript
export interface IExampleReactComponentProps {
    context: BaseComponentContext;
}

return (
  <PeoplePicker context={props.context} ... >
)
```

**Reply by Maya-Mostafa** [reply]
That's brilliant! Thanks @patrikhellgren for your help and detailed explanation. It is finally working for me :)

**Reply by wobba** [reply]
@patrikhellgren want to PR a tutorial/sample on this? 🙂

**Reply by patrikhellgren** [reply]
@wobba Sure I will have a look at that. I'll submit a PR to [microsoft-search/pnp-modern-search-extensibility-samples](https://github.com/microsoft-search/pnp-modern-search-extensibility-samples) later this week or something.

**Comment by MarkusLanger** [comment]
Thanks for the quick response. I will try it out.
